### PR TITLE
Job CLI deploy/list/export/remove

### PR DESCRIPTION
**Pull request by stinkyfingers**
# Description

Adds CLI commands `job deploy`, `job list`, `job export`, `job remove`. Job deploy only implements yaml file deployments. [Spec here](https://docs.google.com/document/d/1fMaLn7ViztU4L8nxV1sTFwLW3-64mWraOiOAE5MnlOw/edit#heading=h.6qrq7urqg756). Makes version, policy, and description fields optional. 

I also found some small fixes that were needed for the Job CRD, noted in comments.

Try it:
`make install ketch manager`
`./bin/manager --enable-leader-election=false --disable-webhooks` (assuming you're not deploying)
`./bin/ketch framework add fw`
`./bin/ketch job deploy job.yaml`
`k get jobs -n ketch-fw`
`k get pods -n ketch-fw`
`k logs <pod> - n ketch-fw pi` 
`./bin/ketch job export myjob`
`./bin/ketch job list`
`./bin/ketch job remove myjob`

example *job.yaml*:
```
name: myjob
version: v1
type: Job
framework: fw
description: "cli test job"
containers:
  - name: pi
    image: perl
    command:
      - "perl"
      - "-Mbignum=bpi"
      - "-wle"
      - "print bpi(2000)"
```

Fixes # 1542

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Chore (documentation addition or typo, file relocation)

## Testing

- [x] New tests were added with this PR that prove my fix is effective or that my feature works (describe below this bullet)
- [ ] This change requires no testing (i.e. documentation update)

## Documentation

- [x] All added public packages, funcs, and types have been documented with doc comments
- [x] I have commented my code, particularly in hard-to-understand areas

## Final Checklist:

- [x] I followed standard [GitHub flow](https://guides.github.com/introduction/flow/) guidelines
- [x] I have performed a self-review of my own code
- [x] My changes generate no new warnings

## Additional Information (omit if empty)

Please include anything else relevant to this PR that may be useful to know.

### CODE REVIEW

File: config/crd/kustomization.yaml [hunk 1]
```diff
@@ -4,6 +4,7 @@
 resources:
 - bases/theketch.io_apps.yaml
 - bases/theketch.io_frameworks.yaml
+- bases/theketch.io_jobs.yaml
```
Comment (stinkyfingers) [inline config/crd/kustomization.yaml:7]: Forgot to add this during the job CRD pr. Makes `make install` install job crd.

File: internal/api/v1beta1/job_types.go [hunk 2]
```diff
@@ -18,23 +17,24 @@ package v1beta1
...
 	"github.com/shipa-corp/ketch/internal/templates"
+
 	v1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
 // JobSpec defines the desired state of Job
 type JobSpec struct {
-	Version      string      `json:"version"`
+	Version      string      `json:"version,omitempty"`
```
Comment (stinkyfingers) [inline internal/api/v1beta1/job_types.go:27]: A few more potentially empty fields.

File: internal/api/v1beta1/job_types.go [hunk 3]
```diff
@@ -45,6 +45,8 @@ type JobStatus struct {
 
 //+kubebuilder:object:root=true
 //+kubebuilder:subresource:status
+// +kubebuilder:resource:scope=Cluster
```
Comment (stinkyfingers) [inline internal/api/v1beta1/job_types.go:48]: I think job scope should be cluster level, like app. Missed this in in the job CRD PR.

File: internal/templates/job/yamls/job.yaml [hunk 4]
```diff
@@ -7,16 +7,16 @@ metadata:
   name: {{ $.Values.job.name }}
 spec:
   {{- if $.Values.job.parallelism }}
-{{ $.Values.job.parallelism | toYaml | indent 4 }}
+  parallelism: {{ $.Values.job.parallelism }}
```
Comment (stinkyfingers) [inline internal/templates/job/yamls/job.yaml:10]: I botched a few template fields in the job CRD PR #116. Fixed here.

File: cmd/ketch/job_deploy.go [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+	if err != nil {
+		return err
+	}
+	setJobSpecDefaults(&spec)
+	if err = validateJobSpec(&spec); err != nil {
+		return err
+	}
+
+	var job ketchv1.Job
+	err = cfg.Client().Get(ctx, types.NamespacedName{Name: spec.Name}, &job)
```
Comment (aleksej-paschenko) [inline cmd/ketch/job_deploy.go:59]: how about using
```golang
import (
	ctrl "sigs.k8s.io/controller-runtime"
)

ctrl.CreateOrUdate()
```

https://github.com/kubernetes-sigs/controller-runtime/blob/master/alias.go#L115
?

File: cmd/ketch/job_deploy_test.go [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+				defer os.Remove(file.Name())
+				tt.filename = file.Name()
+			}
+			out := &bytes.Buffer{}
+			err := jobDeploy(context.Background(), tt.cfg, tt.filename, out)
+			if len(tt.wantErr) > 0 {
+				require.NotNil(t, err)
+				require.Equal(t, tt.wantErr, err.Error())
+				return
+			} else {
```
Comment (aleksej-paschenko) [inline cmd/ketch/job_deploy_test.go:106]: no need to use `else` statement.

File: cmd/ketch/job_export.go [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+	cmd.Flags().StringVarP(&options.filename, "filename", "f", "", "filename for job export")
+	return cmd
+}
+
+func jobExport(ctx context.Context, cfg config, options jobExportOptions, out io.Writer) error {
+	var job ketchv1.Job
+	if err := cfg.Client().Get(ctx, types.NamespacedName{Name: options.name}, &job); err != nil {
+		return err
+	}
+	if options.filename != "" {
```
Comment (aleksej-paschenko) [inline cmd/ketch/job_export.go:52]: we use this pattern multiple times, how about having an independent function for `exporting`?

File: cmd/ketch/job_list.go [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+			Description: item.Spec.Description,
+		})
+	}
+	return output
+}
+
+func jobListNames(cfg config, nameFilter ...string) ([]string, error) {
+	jobs := ketchv1.JobList{}
+	if err := cfg.Client().List(context.TODO(), &jobs); err != nil {
+		return nil, fmt.Errorf("failed to list apps: %w", err)
```
Comment (aleksej-paschenko) [inline cmd/ketch/job_list.go:62]: should it be `failed to list jobs`?

File: cmd/ketch/job_remove_test.go [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			out := &bytes.Buffer{}
+			err := jobRemove(context.Background(), tt.cfg, tt.jobName, out)
+			if len(tt.wantErr) > 0 {
+				require.NotNil(t, err)
+				require.Equal(t, tt.wantErr, err.Error())
+				return
+			} else {
```
Comment (aleksej-paschenko) [inline cmd/ketch/job_remove_test.go:75]: no need to have `else` statement

File: cli_tests/job.sh [hunk 10]
```diff
@@ -0,0 +1,90 @@
+#!/usr/bin/env bats
+
+# To run locally:
+# export KETCH_EXECUTABLE_PATH=<location of ketch binary>
+# assure you have a kubernetes cluster running w/ traefik, cert manager, etc. (see ketch getting started docs)
```
Comment (aleksej-paschenko) [inline cli_tests/job.sh:5]: :+1:

File: cmd/ketch/job_export_test.go [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			out := &bytes.Buffer{}
+			err := jobExport(context.Background(), tt.cfg, tt.options, out)
+			if len(tt.wantErr) > 0 {
+				require.NotNil(t, err)
+				require.Equal(t, tt.wantErr, err.Error())
+				return
+			} else {
```
Comment (aleksej-paschenko) [inline cmd/ketch/job_export_test.go:95]: no need to use `else`

**Review by aleksej-paschenko — COMMENTED** [review]
looks awesome

File: cmd/ketch/job_deploy.go (hunk 5 above)
Comment (stinkyfingers) [reply cmd/ketch/job_deploy.go]: Cool. I didn't even know it existed...until now.

File: cmd/ketch/job_deploy_test.go (hunk 6 above)
Comment (stinkyfingers) [reply cmd/ketch/job_deploy_test.go]: Oops. Thanks for catching all these.

File: cmd/ketch/job_export.go (hunk 7 above)
Comment (stinkyfingers) [reply cmd/ketch/job_export.go:52]: Good call. Refactored into the cmd/output package for the 3 `export` commands to use.

File: cmd/ketch/job_list.go (hunk 8 above)
Comment (stinkyfingers) [reply cmd/ketch/job_list.go]: Good catch. It's possible I was copy/pasting.

File: cmd/ketch/job_export.go (hunk 7 above)
Comment (aleksej-paschenko) [reply cmd/ketch/job_export.go:52]: I talked about the whole piece of code 
```golang
       if options.filename != "" {
		f, err := output.GetOutputFile(options.filename)
		if err != nil {
			return err
		}
		defer f.Close()
		out = f
	}
	b, err := yaml.Marshal(job.Spec)
	if err != nil {
		return err
	}
	_, err = out.Write(b)
```
but it doesn't really matter, it is up to you

**Review by aleksej-paschenko — APPROVED** [review]

**Comment by stinkyfingers** [comment]
Haha @aleksej-paschenko. Got it. It went right over my head. Refactored the whole block.

**Pull request merged**